wit/window.wit

```diff
@@ -0,0 +1,228 @@
+package wasi:webgpu;
+
+interface window {
+    use graphics-context.{graphics-context};
+    use wasi:io/poll@0.2.0.{pollable};
+    
+    
+    /// https://developer.mozilla.org/en-US/docs/Web/API/UI_Events/Keyboard_event_code_values
+    enum key-code {
+        // TODO is this needed? I don't think the MVP needs key codes.
+        unidentified,
+    }
+    
+    /// https://developer.mozilla.org/en-US/docs/Web/API/UI_Events/Keyboard_event_key_values#document_keys
+    /// Excerpt of the most useful keys.
+    enum special-key {
+       alt,
+       alt-graph,
+       caps-lock,
+       ctrl,
+       meta,
+       num-lock,
+       scroll-lock,
+       shift,
+       
+       arrow-down,
+       arrow-left,
+       arrow-right,
+       arrow-up,
+       end,
+       home,
+       page-down,
+       page-up,
+       
+       backspace,
+       clear,
+       delete,
+       insert,
+       paste,
+       
+       context-menu,
+       escape,
+       pause,
+       
+       brightness-down,
+       brightness-up,
+       print,
+       
+       f1,
+       f2,
+       f3,
+       f4,
+       f5,
+       f6,
+       f7,
+       f8,
+       f9,
+       f10,
+       f11,
+       f12,
+       f13,
+       f14,
+       f15,
+       f16,
+       f17,
+       f18,
+       f19,
+       f20,
+       
+       app-switch,
+       
+       channel-down,
+       channel-up,
+       fast-forward,
+       media-pause,
+       media-play,
+       media-play-pause,
+       media-record,
+       media-rewind,
+       media-stop,
+       media-track-next,
+       media-track-previous,
+       
+       audio-volume-down,
+       audio-volume-up,
+       audio-volume-mute,
+    }
+    
+    flags modifiers {
+        ctrl,
+        alt,
+        alt-graph,
+        shift,
+        caps-lock,
+        meta,
+    }
+    
+    /// Describes the key location on the keyboard, for keys that appear more than once on a keyboard.
+    enum key-location {
+        standard,
+        left,
+        right,
+        numpad,
+    }
+    
+    variant key {
+        /// The key has an associated Unicode scalar value.
+        character(u32),
+        /// The key is a key without a Unicode scalar value.
+        special(special-key),
+        /// A dead key was pressed. The next key event will contain the combined character.
+        dead,
+        /// The key could not be identified.
+        unidentified,
+    }
+    
+    
+    record key-data {
+        modifiers: modifiers,
+        key: key,
+        code: key-code,
+        location: key-location,
+        repeat: bool,
+    }
+    
+    flags mouse-buttons {
+        left,
+        middle,
+        right,
+    }
+    
+    record mouse {
+        position: position,
+        buttons: mouse-buttons,
+    }
+    
+    record scroll {
+        mouse: mouse,
```

Comment:
Why does the scroll event need a mouse field, instead of just a position? What does `scroll.mouse.buttons` mean?

Reply:
mouse is the same record as used in the click events, so the client doesn't need to keep track of the mouse state: The scroll event also gets the button state and cursor position, and it has an x and y scroll, though I don't know how many mice with horizontal scrolling there are. That is present on the web, but the Web wants compatibility with literally everything, so that doesn't mean much. The web wheel event even has z scrolling, but I didn't add that.

Reply:
Oh! So you get the full mouse state on each event. Makes sense.